### PR TITLE
NAPPS-1458: possible more permanent fix for healthcheck failure in deployed app

### PR DESCRIPTION
**Pull request by Kat-Alo**
I don't like this solution, but I thought it'd be quick to demonstrate one possibility by opening a PR so we can discuss Monday. 

Context: we are redirecting requests made to HTTP:80 to HTTPS:443 because we set `http_to_https_redirect` to `true`. From my limited understanding, this is best security practice. As a result, however, our health check is returning 300+ status codes to indicate there is a redirect. We temporarily expanded our accepted "success" codes to include 300-399, but my concern is that 500 health check status codes would also surface as 300 — reducing the meaningfulness of our health check. 

(As an aside: I don't understand how the health check could be active, but the app down. In other words, I don't understand the context in which the health check could return a 500, but I have to assume that's possible. Otherwise, what is the point of the health check, right? Could it be for database connection errors?)

After going through [our StackJack ECS template](https://github.com/WPMedia/aws-templates/blob/78cde3a8ff9b5b14c11fc6f924fb86977ebffa8d/templates/v1/cfn/shared/apps/ecs-v2/service.template.yml#L357-L370) (and a couple quick searches through the rest of the templates), it seems like there isn't a way to set the health check protocol and port such that a redirect won't be triggered, even though [AWS CloudFormation docs](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html#health-check-settings) suggest it's possible to set `HealthCheckPort` and `HealthCheckProtocol`.

As such, I've pulled the template into our repo and modified it in the way _I believe_ would work, but it's not ideal. This seems like an issue that other teams would have to face as well, so I would prefer to open a PR against the StackJack template PR if we think we can get an approval. And, in the meantime, we can accept status codes 200-399, as we are now. Would love to hear others' thoughts though.

[NAPPS-1458](https://arcpublishing.atlassian.net/browse/NAPPS-1458?atlOrigin=eyJpIjoiODcyNDQ3YWRhZmM2NGJkMmI2MjdkMDgzZGU4YzBhY2YiLCJwIjoiaiJ9)

[NAPPS-1458]: https://arcpublishing.atlassian.net/browse/NAPPS-1458?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test.yml [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+        Type: AWS::ElasticLoadBalancingV2::TargetGroup
+        Properties:
+
+            VpcId: !ImportValue "vpc-dev-VpcId"
+            Port: 3001
+            Protocol: HTTP
+            HealthCheckPath: /healthcheck
+            Matcher: {'HttpCode': '200-299'}
+            HealthCheckIntervalSeconds: 10
+            HealthyThresholdCount: 2
```
Comment (aaronbrezel) [inline test.yml]: 🤔 If we can manage to set `HealthCheckProtocol` to `HTTPS` as detailed in [this documentation](
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html#cfn-elasticloadbalancingv2-targetgroup-healthcheckprotocol), I think we will avoid the redirect.

File: test.yml [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+    DefaultTargetGroupHttp:
+        
+        Metadata: {'Stackjack': {'AllowDestroy': true } }
+        Type: AWS::ElasticLoadBalancingV2::TargetGroup
+        Properties:
+
+            VpcId: !ImportValue "vpc-dev-VpcId"
+            Port: 3001
+            Protocol: HTTP
```
Comment (aaronbrezel) [inline test.yml]: 🤔 I am not 100% sure what the purpose of [Protocol](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html#cfn-elasticloadbalancingv2-targetgroup-protocol) is. It seems to determine what protocol is used when load balancing to different targers (servers?).

That being said, to remove our specific issue, I think we are more concerned with `HealthCheckProtocol`.

File: test.yml [hunk 3]
```diff
@@ -0,0 +1,470 @@
+
```
Comment (aaronbrezel) [inline test.yml]: I am pretty sure we'll want to 🗑️ this file once we're ready to merge, right?

File: cfn/configs/klaxon/dev/us-east-1/config.yml [hunk 4]
```diff
@@ -332,7 +332,9 @@ context:
         healthy_threshold_count: 2
         unhealthy_threshold_count: 6
         matcher:
-          HttpCode: "200-399" # accepts success or redirect
+          HttpCode: "200-299"
```
Comment (aaronbrezel) [inline cfn/configs/klaxon/dev/us-east-1/config.yml:335]: 💭 Back to the old value

File: cfn/templates/service.template.yml [hunk 5]
```diff
@@ -0,0 +1,1754 @@
...
+            router.health_check.matcher | default(service.health_check_matcher) | default({"HttpCode": "200-299"})) }}
+        {{- update_element(router.health_check, "interval_seconds",
+            router.health_check.interval_seconds | default(service.health_check_interval_seconds) | default(10)) }}
+        {{- update_element(router.health_check, "healthy_threshold_count",
+            router.health_check.healthy_threshold_count | default(service.healthy_threshold_count) | default(2)) }}
+        {{- update_element(router.health_check, "unhealthy_threshold_count",
+            router.health_check.unhealthy_threshold_count | default(service.unhealthy_threshold_count) | default(6)) }}
+        {{- update_element(router.health_check, "timeout_seconds",
+            router.health_check.timeout_seconds | default(service.health_check_timeout_seconds) | default(5)) }}
+        {{- update_element(router.health_check, "protocol",
```
Comment (aaronbrezel) [inline cfn/templates/service.template.yml:370]: 💭 Just to make sure I've got the scope of the changes here. We're adding an additional `protocol` and `port` here.

File: cfn/templates/service.template.yml [hunk 6]
```diff
@@ -0,0 +1,1754 @@
...
+            Port: {{ router.container_port }}
+            Protocol: {{ router.container_protocol | upper }}
+
+          {%- if router.perform_health_check %}
+            HealthCheckPath: {{ router.health_check.path }}
+            Matcher: {{ router.health_check.matcher }}
+            HealthCheckIntervalSeconds: {{ router.health_check.interval_seconds }}
+            HealthyThresholdCount: {{ router.health_check.healthy_threshold_count }}
+            UnhealthyThresholdCount: {{ router.health_check.unhealthy_threshold_count }}
+            HealthCheckTimeoutSeconds: {{ router.health_check.timeout_seconds }}
```
Comment (aaronbrezel) [inline cfn/templates/service.template.yml:1166]: In order for the deployed `AWS::ElasticLoadBalancingV2::TargetGroup` to be altered, I think we would need to also change this section of code, as this is the section of code which actually determines the config of the resource.

**Review by aaronbrezel — COMMENTED** [review]
I think this is the right track! Makes sense to me, especially considering the [AWS resource documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html#cfn-elasticloadbalancingv2-targetgroup-healthcheckprotocol) specifically states how we can set the healthcheck protocol. 

I agree that this might be a good opportunity to practice contributing directly to the [aws-templates](https://github.com/WPMedia/aws-templates/blob/master/templates/v1/cfn/shared/apps/ecs-v2/service.template.yml). Should we keep this PR as a guide and open up an actual PR in that repo?

Lastly, are there a few files in this PR that we could delete? Namely anything that's not the `cfn/configs/klaxon/dev/us-east-1/config.yml` and `cfn/templates/service.template.yml`?

File: cfn/configs/klaxon/dev/us-east-1/config.yml (hunk 4 above)
Comment (Kat-Alo) [reply cfn/configs/klaxon/dev/us-east-1/config.yml:335]: yes! the hope would be that this fix means we'd no longer anticipate a 300 code

**Comment by Kat-Alo** [comment]
> Lastly, are there a few files in this PR that we could delete? Namely anything that's not the `cfn/configs/klaxon/dev/us-east-1/config.yml` and `cfn/templates/service.template.yml`?

@aaronbrezel  whoops! that was sloppy. will fix that...

File: cfn/templates/service.template.yml (hunk 5 above)
Comment (Kat-Alo) [reply cfn/templates/service.template.yml:370]: yep!

File: .gitignore [hunk 7]
```diff
@@ -11,4 +11,6 @@ pull_dev_database.sh
 
 snapshots
 .env*
-!.env.*.example
\ No newline at end of file
+!.env.*.example
+
+public/assets/
```
Comment (Kat-Alo) [inline .gitignore:16]: ignoring precompiled assets. The dockerfile runs a command to pre-compile these for the deployed app

File: cfn/configs/klaxon/dev/us-east-1/config.yml [hunk 8]
```diff
@@ -27,7 +27,7 @@ _globals:
   - !_join &CLUSTER_NAME ["-", *CLUSTER_PREFIX, *ENVIRONMENT]
   - !_join &STACK_NAME ["-", *CLUSTER_NAME, *SERVICE_NAME]
 
-template: v1/cfn/shared/apps/ecs-v2/service.template.yml
+template: cfn/templates/service.template.yml
```
Comment (Kat-Alo) [inline cfn/configs/klaxon/dev/us-east-1/config.yml:30]: confirmed with stackjack build that this is the correct way to reference a local template

File: cfn/configs/klaxon/dev/us-east-1/config.yml [hunk 9]
```diff
@@ -327,12 +327,14 @@ context:
       #
       health_check:
         path: "/healthcheck"
-        interval_seconds: 10
+        interval_seconds: 30
```
Comment (Kat-Alo) [inline cfn/configs/klaxon/dev/us-east-1/config.yml:330]: raising this interval per aws default for https health checks

File: test.yml (hunk 1 above)
Comment (Kat-Alo) [reply test.yml]: yes! this is exactly the hope

**Review by emlorraine — APPROVED** [review]
👩‍⚕️

**Pull request merged**